<launch>
    <arg name="model_name"/>
    <!-- Robot Localization nodes-->

    <!-- Continuous UKF -->
    <!-- Used for navigation, provides continuous information -->
    <!-- Transforms from odom to base_link using TurtleBot's self odometry and IMU -->
    <node pkg="robot_localization" type="ukf_localization_node" name="ukf_continuous" clear_params="true" output="screen">
        <remap from="odometry/filtered" to="odometry/filtered_continuous"/>
        <remap from="set_pose" to="set_pose_continuous"/>
        <param name="frequency" value="10"/>
        <param name="sensor_timeout"  value="0.1"/>
        <param name="two_d_mode" value="true"/>
        <param name="odom_frame" value="$(arg model_name)/odom"/>
        <param name="base_link_frame" value="$(arg model_name)/base_footprint_filter"/>
        <param name="world_frame" value="$(arg model_name)/odom"/>
        <param name="transform_time_offset" value="0.0"/>
        <param name="odom0" value="odom_remapped"/> <!-- internal odom-->
        <rosparam param="odom0_config">[false,  false,  false, <!-- x y z -->
                                        false,  false,  false, <!-- roll pitch yaw -->
                                        true,   true,   false, <!-- x_vel y_vel z_vel -->
                                        false,  false,  true, <!-- roll_vel pitch_vel yaw_vel -->
                                        false,  false,  false] <!-- x_acc y_acc z_acc -->
        </rosparam>
        <param name="imu0" value="imu_data_remapped"/>
        <!-- TODO reinclude x_acc once using real turtlebot and see if it has a better IMU than Gazebo -->
        <!-- Because Gazebo produces constant x_acc of .2-.3 -->
        <rosparam param="imu0_config">[false,   false,  false, <!-- x y z -->
                                    false,   false,  true, <!-- roll pitch yaw -->
                                    false,   false,  false, <!-- x_vel y_vel z_vel -->
                                    false,   false,  true, <!-- roll_vel pitch_vel yaw_vel -->
                                    false,    false,  false] <!-- x_acc y_acc z_acc -->
        </rosparam>
        <param name="odom0_differential" value="false"/>
        <param name="imu0_differential" value="false"/>
        <param name="odom0_relative" value="false"/>
        <param name="imu0_relative" value="false"/>
        <param name="print_diagnostics" value="true"/>
    </node>

    <!-- Discrete UKF -->
    <!-- Transforms from map to odom using continuous filter output and external pose readings -->
    <node pkg="robot_localization" type="ukf_localization_node" name="ukf_discrete" clear_params="true" output="screen">
        <remap from="odometry/filtered" to="odometry/filtered_discrete"/>
        <remap from="set_pose" to="set_pose_discrete"/>
        <param name="frequency" value="10"/>
        <param name="sensor_timeout"  value="0.1"/>
        <param name="two_d_mode" value="true"/>
        <param name="odom_frame" value="$(arg model_name)/odom"/>
        <param name="base_link_frame" value="$(arg model_name)/base_footprint_filter"/>
        <param name="world_frame" value="map"/>
        <param name="map_frame" value="map"/>
        <param name="transform_time_offset" value="0.0"/>
        <param name="odom0" value="odometry/filtered_continuous"/> <!-- odom output from continuous filter-->
        <!-- TODO adjust these -->
        <rosparam param="odom0_config">[false,   false,   false, <!-- x y z -->
                                        false,  false,  true, <!-- roll pitch yaw -->
                                        true,   true,  false, <!-- x_vel y_vel z_vel -->
                                        false,  false,  true, <!-- roll_vel pitch_vel yaw_vel -->
                                        true,  false,  false] <!-- x_acc y_acc z_acc -->
        </rosparam>
        <param name="pose0" value="external_poses"/>
        <rosparam param="pose0_config">[true,   true,  false, <!-- x y z -->
                                    false,   false,  false, <!-- roll pitch yaw -->
                                    false,   false,  false, <!-- x_vel y_vel z_vel -->
                                    false,   false,  false, <!-- roll_vel pitch_vel yaw_vel -->
                                    false,    false,  false] <!-- x_acc y_acc z_acc -->
        </rosparam>
        <param name="odom0_differential" value="false"/>
        <param name="pose0_differential" value="false"/>
        <param name="odom0_relative" value="false"/>
        <param name="pose0_relative" value="false"/>
        <param name="print_diagnostics" value="true"/>
    </node>

    <node pkg="tf" type="static_transform_publisher" name="filter_footprint_to_link_tf" args="0 0 0.01 0 0 0 1 $(arg model_name)/base_footprint_filter $(arg model_name)/base_link_filter 50"/>

    <node pkg="cwru_turtlebot" type="sensor_remap.py" name="sensor_remap"/>
</launch>